### ISSUE TITLE
Add a basic testing framework to the CLI

### ISSUE THREAD

**Issue by LaurenzV** [typst/typst#2381]
### Description

It would be nice if there were a command such as `typst test` that allows to run visual regression tests for a package, similar to `cargo test`. The main motivation for this is that I saw some packages such as `tablex` or `cetz` have their own implementations of regression tests, so I think it would be great if we could unify this in some way so that not every package creator has to roll their own testing system.

I think a good start would be to just include visual regression tests (since this shouldn't be that difficult to implement either), and at a later state we can also think about how Typst code itself could be tested as well.

### Use Case

See above.

**Comment by johannes-wolf** [comment]
For reference: Cetz uses a shell-script https://github.com/johannes-wolf/cetz/blob/master/scripts/test which looks for tests (`test.typ` files), compiles them to PNG and compares them against a reference image (`ref.png`, next to the test file) using ImageMagick. On failure, a difference image is stored next to the test.

**Comment by LaurenzV** [comment]
I think we wouldn't even need something like ImageMagick, we can just use Typst's native PNG export. This does mean that SVG and PDF won't be tested, but from the perspective of a package author this is irrelevant, since if there are discrepancies between the output formats it's upon us to fix this, and not the author.

**Comment by laurmaedje** [comment]
As I understand it, cetz's tests do use PNG export. ImageMagick is only used for the comparison.

**Comment by LaurenzV** [comment]
Ah, right. But I think we can just use a pixel-comparison based approach, similar to how it's done in svg2pdf, and create diff images in case there are any discrepancies.

**Comment by LaurenzV** [comment]
So, here are is a first suggestion for how it might look like:

We introduce a new sub command `typst test` that allows to test a package. It provides the following options:
# Running tests
In order to run the tests, you can in principle just run the command above. What will happen is that it will go through all of the test files, generate an image from them and compare them against the reference image. If they match, the test passes, if not it fails. If no corresponding reference image was found, the test will be marked as skipped. If a test fails, a diff image will be generated that can be inspected.

In addition, the following parameters are supported:
`--font-path`: Provide a path to a directory of fonts that should be included. `typst test` will by default only use the fonts embedded in the CLI but not system fonts, so as to ensure reproducibility across different platforms, for example in a GitHub CLI.
`--exact`: Only run the tests whose filename matches the pattern.
`--tests-dir`: The path to the folder containing the tests. By default 'tests'

# Reference images
`--update`: Updates all currently existing reference images. You can additionally pass a pattern to only match files that match the pattern.
`--generate`: Generates reference images for all test files that don't have any yet. As above, you can additionally pass a pattern.

# Folder structure
This is how the folder structure could look like:
```
.
└── MyPackage/
    ├── main.typ
    ├── helper.typ
    └── tests/
        ├── assets/
        │   └── some_stuff.json
        ├── ref/
        │   ├── numeric/
        │   │   ├── addition.png
        │   │   └── subtraction.png
        │   └── text/
        │       └── format.png
        ├── src/
        │   ├── numeric/
        │   │   ├── addition.typ
        │   │   └── subtraction.typ
        │   └── text/
        │       └── format.typ
        └── diff/
            └── numeric/
                └── addition.png
```

Inside of the `tests` folder, there needs to be another `src` folder which contains all of the test cases. Reference images will be created by the program inside a `ref` folder, diff images inside of the `diff` folder.

The root folder will be wherever `typst test` was invoked, so the intended usage is that you would invoke `typst test` inside `MyPackage` and then can just do `#import "/main.typ` in each test file to import the actual package. If you want to include assets, you could place them inside of the `test` folder as well and then include them by something like `../../assets/some_stuff.json`.

# Discussion points
- Something I don't like about the approach mentioned above is that the folder structure is still kind of rigid, i.e. the user has to put their tests inside of the `src` folder for them to be detected. I guess we could also change it so that _any_ `.typ` files found in the `tests` folder will be considered, but then it's a bit hard to figure out where to store the reference images. Should they be stored inside of the same folder as the original `.typ` file instead of a completely separate `ref` that mirrors the directory structure?
- I think in the ideal case, we could make use of `typst.toml` in the future to allow configuring such settings inside of that file instead of having to provide them all as arguments in the CLI. However, since Typst doesn't have a notion of a "project" (yet?), we probably can't really make use of that now.

Any ideas? @andreasKroepelin @PgBiel @johannes-wolf  since you all have bigger libraries that probably could profit from such a feature, feel free (but not obliged!) to let me know what you think of that proposal.

**Comment by andreasKroepelin** [comment]
I like that proposal a lot! What exactly would a diff image be? I have seen interactive diff viewers for images but how would one show the diff in a static image?

**Comment by LaurenzV** [comment]
> I like that proposal a lot! What exactly would a diff image be? I have seen interactive diff viewers for images but how would one show the diff in a static image?
Basically something like this:
<img width="694" alt="image" src="https://github.com/typst/typst/assets/47084093/ee74ac94-0d4e-48e9-9237-5ac982879790">

**Comment by johannes-wolf** [comment]
@LaurenzV The only advantage of having Typst becoming a test-runner is that it could use a “warmed up” compiler with memoization shared across tests.

I think running tests can be handled better by external scripts, because implementing all this into Typst + running tests in parallel + image diffing + … might be a bit of a stretch.
Other tools can do this perfectly fine, atm. GNU parallel + ImageMagick, …
Re-Implementing such functionality insides Typst seems less flexible to me. I suggest having an external project that provides some scripts for handling tests instead of having this in the Typst binary.

**Comment by LaurenzV** [comment]
Fair enough! Although I personally do think that there is a lot of value in having a standardized and streamlined way of testing your packages integrated into Typst. It would be much more convenient if it would _just work_ with the existing binary instead of having to install yet another program. I also don't think it needs to be anything fancy, e.g. image diffing can be implemented pretty easily, similar to how we have it in `svg2pdf`.

But I do agree that I'm also not sure whether it should really be part of the `typst` binary, maybe in the long run there should be a separate program similar to `cargo` that contains additional modules (such as testing, formatting, etc.). Not sure. I guess in the end it's up to @laurmaedje to decide.

**Comment by andreasKroepelin** [comment]
> Basically something like this:
> <img width="694" alt="image" src="https://github.com/typst/typst/assets/47084093/ee74ac94-0d4e-48e9-9237-5ac982879790">
> 

What am I seeing here? 😅

**Comment by LaurenzV** [comment]
> What am I seeing here? 😅

On the very left you see the reference image, on the very right the actual output, and in the middle the difference between the two.

**Comment by laurmaedje** [comment]
> But I do agree that I'm also not sure whether it should really be part of the typst binary, maybe in the long run there should be a separate program similar to cargo that contains additional modules (such as testing, formatting, etc.).

Just wanted to quickly comment on this: In general, I wouldn't want a second official binary. I don't see the value in splitting things up across two binaries.

**Comment by LaurenzV** [comment]
> Oh, I assume those are red and green areas then? Sorry, those are often difficult to distinguish for me 😅

Oh yeah, sorry, maybe the example was chosen badly. 😅 Basically on the left you can see a green SVG image, on the right you can see the same one but slightly different, and in the middle you see red lines that indicate where the images on the left and right differ.

**Comment by LaurenzV** [comment]
But I'm also starting to think that maybe for now I will start a separate project (similar to `typstfmt`) as a kind of playground to see what's possible and what makes sense, and maybe at some point it can then be merged into the official binary.

**Comment by LaurenzV** [comment]
Will close this, since there is [typst-test](https://github.com/tingerrr/typst-test) now and I'm not convinced anymore that this should be part of the CLI.
